src/index.js

```diff
@@ -1,174 +1,226 @@
-   * @param {Function} fn
+   * @param {String} room - the room to get the clients connected to
+   * @return {Promise.<String[]>} containing room names
    * @api public
    */
 
-  function clients(room, fn = noop) {
-    room = rooms[room] || {}
-    setImmediate(fn, null, keys(room))
+  function clients(room) {
+    const socketSet = rooms.get(room);
+    if (socketSet) {
+      return Promise.resolve(Array.from(socketSet));
+    }
+    return Promise.resolve([]);
   }
 
   /**
    * Remove all sockets from a room.
    *
-   * @param {String|Array} room
-   * @param {Function} [fn]
+   * @param {String|String[]} room - the room(s) to empty
+   * @return {Promise} no data
    * @api public
    */
 
-  function empty(room, fn = noop) {
-    if (room && rooms[room]) {
-      if (!isArray(room)) clear(room)
-      else for (const rm of room) clear(rm)
+  function empty(room) {
+    if (isArray(room)) {
+      room.map(rm => clear(rm));
+      return Promise.all(room);
     }
-    setImmediate(fn, null, null)
+    return clear(room);
   }
 
   /**
    * Reset all rooms or a particular room.
    *
-   * @param {String} room
-   * @param {Function} [fn]
+   * @param {String} room - the room to clear
+   * @return {Promise} no data
    * @api public
    */
-
-  function clear(room, fn = noop) {
-    if ('function' === typeof room) {
-      rooms = {}, sids = {}
+  function clear(room) {
+    if (!room) {
+      rooms.clear();
+      sids.clear();
+      return Promise.resolve();
     }
-    for (const id in rooms[room]) {
-      delete sids[id][room]
+    const sockets = rooms.get(room);
+    if (sockets) {
+      for (const id of sockets) {
+        const socketRooms = sids.get(id);
+        socketRooms.delete(room);
+        if (socketRooms.size === 0) {
+          sids.delete(id);
+        }
+        rooms.delete(room);
+        // TODO update wild 
+        // wild.remove(room)
+      }
     }
-    delete rooms[room]
-    wild.remove(room)
-    setImmediate(fn, null, null)
+    return Promise.resolve();
   }
 
   /**
    * Broadcast a packet.
    *
-   * Options:
-   *  - `except` {Array} sids that should be excluded
-   *  - `rooms` {Array} list of rooms to broadcast to
-   *  - `method` {String} 'write' or 'send' if primus-emitter is present
-   *
-   * @param {Object} data
-   * @param {Object} opts
-   * @param {Object} clients Connected clients
+   * @param {*} data - the data to transmit
+   * @param {Object} options - options to use
+   * @param {String[]} options.except - spark IDs to exclude
+   * @param {String[]} options.rooms - list of the rooms to broadcast to
+   * @param {String} options.method - 'write' or 'send' if primus-emitter is present
+   * @param {Function} options.transformer - the transformer function to call on the data
+   * @param {Object} clients -  Connected clients
+   * @return {Promise.<Set>} set of sparkIDs transmitted to
    * @api public
    */
-
+  
   function broadcast(data, options = {}, clients) {
```

Comment:
There is a problem in this function.
The original version used the `sent` object to prevent data from being sent to same socket more than once. This can happens for example when a socket joined more rooms and we are broadcasting to those specific rooms.

The socket should receive the message only once.